### ISSUE TITLE
BBC: Blead breaks OVID/Acme-Dot-1.10.tar.gz

### ISSUE THREAD

**Issue by andk** [Perl/perl5#19460]
**Description**

My smokers have stopped producing PASS reports for both
- OVID/Acme-Dot-1.10.tar.gz and
- DAVECROSS/Tie-Hash-Cannabinol-v1.12.2.tar.gz

Bisect leads to
``` text
f6387cff9cb31db4cf18c8641917ea4639ac2b65 is the first bad commit
commit f6387cff9cb31db4cf18c8641917ea4639ac2b65
Author: Yves Orton <demerphq@gmail.com>
Date:   Mon Feb 14 07:34:54 2022 +0100

    Reapply squashed "avoid identical stack traces" patches
```

**Steps to Reproduce**

cpan -i Acme::Dot

or

cpan -i Tie::Hash::Cannabinol

**Expected behavior**

tests should pass and modules should be installed

**Perl configuration**

```
Summary of my perl5 (revision 5 version 35 subversion 9) configuration:
  Commit id: f6387cff9cb31db4cf18c8641917ea4639ac2b65
  Platform:
    osname=linux
    osvers=5.15.0-2-amd64
    archname=x86_64-linux-thread-multi-ld
    uname='linux k93msid 5.15.0-2-amd64 #1 smp debian 5.15.5-2 (2021-12-18) x86_64 gnulinux '
    config_args='-Dprefix=/home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.8-105-gf6387cff9c/8854 -Dmyhostname=k93msid -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Dlibswanted=cl pthread socket inet nsl gdbm dbm malloc dl ld sun m crypt sec util c cposix posix ucb BSD gdbm_compat -Duseithreads -Duselongdouble -DEBUGGING=-g'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=define
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2 -g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='11.2.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='long double'
    nvsize=16
    Off_t='off_t'
    lseeksize=8
    alignbytes=16
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib
    libs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.33.so
    so=so
    useshrplib=false
    libperl=libperl.a
    gnulibc_version='2.33'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -g -L/usr/local/lib -fstack-protector-strong'


Characteristics of this binary (from libperl): 
  Compile-time options:
    HAS_TIMES
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_LONG_DOUBLE
    USE_PERLIO
    USE_PERL_ATOF
    USE_REENTRANT_API
    USE_THREAD_SAFE_LOCALE
  Built under linux
  Compiled at Feb 25 2022 04:35:12
  @INC:
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.8-105-gf6387cff9c/8854/lib/site_perl/5.35.9/x86_64-linux-thread-multi-ld
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.8-105-gf6387cff9c/8854/lib/site_perl/5.35.9
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.8-105-gf6387cff9c/8854/lib/5.35.9/x86_64-linux-thread-multi-ld
    /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.35.8-105-gf6387cff9c/8854/lib/5.35.9


```

**Comment by haarg** [comment]
Tie::Hash::Cannabinol relies on a feature of Attribute::Handlers that was broken before, but is now more broken. I've created a PR trying to address this: #19461

Acme::Dot attempts to search further up the call stack from where its import method is called, but it's not obvious what it expects `caller(1)` to even represent. It seems like it only previously worked by accident.

**Comment by jkeenan** [comment]
commit f6387cff9cb31db4cf18c8641917ea4639ac2b65
Author:     Yves Orton <demerphq@gmail.com>
AuthorDate: Mon Feb 14 07:34:54 2022 +0100
Commit:     Paul "LeoNerd" Evans <leonerd@leonerd.org.uk>
CommitDate: Sat Feb 19 11:23:43 2022 +0000

@demerphq @leonerd , can you take a look?  Thanks.

**Comment by Leont** [comment]
@Ovid 

> Acme::Dot attempts to search further up the call stack from where its import method is called, but it's not obvious what it expects caller(1) to even represent. It seems like it only previously worked by accident.

Yeah, this very much looks like a case of "this only worked accidentally", and given it's an Acme module I'm not sure we should care.

**Comment by demerphq** [comment]
On Fri, 25 Feb 2022 at 10:23, Graham Knop ***@***.***> wrote:

> Tie::Hash::Cannabinol relies on a feature of Attribute::Handlers that was
> broken before, but is now more broken. I've created a PR trying to address
> this: #19461 <https://github.com/Perl/perl5/pull/19461>
>

Merged.


> Acme::Dot attempts to search further up the call stack from where its
> import method is called, but it's not obvious what it expects caller(1)
> to even represent. It seems like it only previously worked by accident.
>
Hopefully Ovid can speak up as to his intent.

cheers,
Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
On Fri, 25 Feb 2022 at 23:19, Leon Timmermans ***@***.***>
wrote:

> @Ovid <https://github.com/Ovid>
>
> Acme::Dot attempts to search further up the call stack from where its
> import method is called, but it's not obvious what it expects caller(1) to
> even represent. It seems like it only previously worked by accident.
>
> Yeah, this very much looks like a case of "this only worked accidentally",
> and given it's an Acme module I'm not sure we should care.
>
Yes, I agree.  The logic doesn't make sense at multiple levels.

It happened to work because the test code executes in main, and caller(1)
in older perls would be erroneously main, and so it injects an AUTOLOAD
method into main.

TO make it work it would have to inject an import method into Foo and then
have that inject an  AUTOLOAD method into its importer. It would also have
to deal with various edge cases of preexisting import methods and
preexisting AUTOLOAD methods. The whole thing gives me the shivers.

I think this module should just be deleted.

Yves


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by jkeenan** [comment]
Tie-Hash-Cannabinol once again installs against blead.  So now this ticket is solely concerned with the failures in Acme-Dot.

I have created the following ticket to alert the maintainer to the problem:
[RT 141733](https://rt.cpan.org/Ticket/Display.html?id=141733).

**Comment by eserte** [comment]
It seems to me that DAVECROSS/Tie-Hash-FixedKeys-v1.13.2.tar.gz and DAVECROSS/Tie-Hash-Regex-1.14.tar.gz are also affected by this problem (both also from DAVECROSS and having the same Attribute::Handlers magic like Tie-Hash-Cannabinol-v1.12.2).

**Comment by haarg** [comment]
Those two dists both seem to pass their tests on blead.

**Comment by rjbs** [comment]
@Ovid (Any response?)

But meantime, I am marking this not blocking v5.36.0

**Comment by Ovid** [comment]
Sorry, I really need to fix my github notifications.

I inherited `Acme::Dot` from Simon Cozens. Never tried to use it myself, even though it seemed fun (and I'd much prefer the dot syntax, even though we're never going to get it).

If anything, I should maybe upload that module again and list it as deprecated, or hand it over to _anyone_ who cares.
